src/backend/distributed/utils/metadata_cache.c

```diff
@@ -480,6 +492,68 @@ ResolveGroupShardPlacement(GroupShardPlacement *groupShardPlacement,
 
 
 /*
+ * LookupNodeForGroup searches the WorkerNodeHash for a worker which is a member of the
+ * given group and also readable (a primary if we're reading from primaries, a secondary
+ * if we're reading from secondaries). If such a node does not exist it emits an
+ * appropriate error message.
+ */
+static WorkerNode *
+LookupNodeForGroup(uint32 groupId)
+{
+	WorkerNode *workerNode = NULL;
+	HASH_SEQ_STATUS status;
+	HTAB *workerNodeHash = GetWorkerNodeHash();
+	bool foundAnyNodes = false;
+
+	hash_seq_init(&status, workerNodeHash);
+
+	while ((workerNode = hash_seq_search(&status)) != NULL)
+	{
```

Comment:
This is no slower than the call to `PrimaryNodeForGroup` which was already there. And it is inefficient to walk through the set of all worker nodes but our biggest customer only has 50 so I don't think this will have a large impact. But, there are a few options:
- Build a second WorkerNodeHash which maps from `groupid` to a list of nodes.
- Store resolved shards in the metadata cache and expire the cache whenever `pg_dist_node` changes. This means we only have to pay the cost on the first query, but has the problem from earlier where we error out if any group is missing a worker, even though we might not need the unresolvable placement.
- Push the resolution deeper into the planner / executors, to make sure we only run resolution on the shards we actually need to run it on.
- Somewhere in `ShardPlacementList` we can build a hash from groupid -> worker node, and pass it into `ResolveGroupShardPlacement`. This won't share between queries but we'll only scan the workernodehash once per query, instead of once per shard, and isn't a lot of extra coding.